### DISCUSSION TITLE
compute_derivative on hills example

### DISCUSSION THREAD

**Discussion by tloom** [pyvista/pyvista#2680] [Q&A]
Hi, apologies if I have misunderstood this but I have been using the compute_derivative function and to understand the outputs better I was trying it with the hills example.

I am currently under the impression that compute_derivative when applied to a vector field returns the partial derivatives in the order du/dx, du/dy, du/dz, dv/dx ... dw/dz. However when I set the vector (u,v,w) to (x,y,z) I would be expecting dx/dx, dy/dy, dz/dz to each be equal to 1. Additionally, when I calculate the divergence I would expect it to be equal to 3 and it is currently equal to 2. 

Below is the example I've been using.
Thank you in advance!

```py
from pyvista import examples
hills = examples.load_random_hills()
 
hills.point_data['xyz']=hills.points
div = hills.compute_derivative('xyz', gradient=True, divergence=True) #default point data

chk=[div['gradient'][0][i] for i in [0,4,8]] #[dx/dx, dy/dy, dz/dz] for first point
chk2=all(abs(div['divergence']-3)<1e-6) #dx/dx+dy/dy+dz/dz=3
```

**Comment by MatthewFlamm** [comment]
When I run this, I get that $\dfrac{dx}{dx}$ and $\dfrac{dy}{dy}$ are both close to 1.  Only $\dfrac{dz}{dz}$ is not.  The hills dataset is a 2D like dataset that is distributed in the `x-y` space but with variation in `z`.  This can also be seen in the divergence which returns 2 for a 2D like dataset like this.  The reason that we do not get identical values to 1 for the $\dfrac{dx}{dx}$ and $\dfrac{dy}{dy}$ ~is that we have numerical error due to the meshing~ is that the mesh isn't planar.

If we use a 3D volume filling grid like `pyvista.UniformGrid` that has regular spacing, we get what you are expecting:

```py
import pyvista
import numpy as np
grid = pyvista.UniformGrid(dims=(11, 11, 11))
grid.point_data['xyz'] = grid.points
div = grid.compute_derivative('xyz',gradient=True, divergence=True)
assert np.allclose(div['gradient'][0][0::4], [1.0, 1.0, 1.0])
assert np.allclose(div['divergence'], [3.0])
```

**Reply by tloom** [reply]
Firstly, thank you very much for your help on this.

We now understand the 3D and 2D planar cases where the gradients and determinants are simpler and thank you for the example because that helped a lot to separate these cases from our own.

We agree that the variation in dz/dz from 0 and the variation in dx/dx and dy/dy from 1 being due to the mesh not being planar and have had a further look into what causes these variations to fully understand the variables, in particular looking at the VTK documentation [https://gitlab.kitware.com/vtk/textbook/raw/master/VTKBook/VTKTextBook.pdf] (pdf pg 302, sec 8.5) but still have a few more questions if you'd be happy to help.

By our current understanding:
The cell types are type 5, ie. triangles, so parameterized with only r and s.
We understand that the compute_gradient method works by firstly computing the gradients in parameter space,
then computing the jacobian between changes in x,y,z and r,s,t and finally using this jacobian to compute the variation in x,y,z

In our case (x,y,z) data away from our surface does not exist, so our surface only exists parametrically in r and s.

Our main question lies in how compute_derivative handles the t component in both the formation and inversion of the Jacobian?
Secondly as the gradient is constrained to the surface, do we expect the divergence to be 2 as it is in essence only a 2D gradient?
